In [8]:
import os
import subprocess

In [6]:
data_path ="/home/mlops-zoomcamp/notebooks/data/"
data_path_output = "/home/mlops-zoomcamp/notebooks/data/output/"

Q1. Install MLflow

In [29]:
os.listdir(data_path_output)

['val.pkl', 'test.pkl', 'dv.pkl', 'train.pkl']

In [15]:
pip_results = subprocess.run(["pip","list"], stderr = subprocess.PIPE, stdout = subprocess.PIPE)
list(filter(lambda p: "mlflow" in p, pip_results.stdout.decode("utf-8").splitlines()))

['mlflow                             2.22.0',
 'mlflow-skinny                      2.22.0']

Q2. How many files were saved to OUTPUT_FOLDER?

*!python  "/home/mlops-zoomcamp/cohorts/2025/02-experiment-tracking/homework/preprocess_data.py" --raw_data_path "/home/mlops-zoomcamp/notebooks/data" --dest_path "/home/mlops-zoomcamp/notebooks/data/output"*

1\
3\
**4**\
7

RMSE: 5.431162180141208


Q3. Train a model with autolog

python /home/mlops-zoomcamp/cohorts/2025/02-experiment-tracking/homework/train.py --data_path /home/mlops-zoomcamp/notebooks/data/output/
rmse:  5.431162180141208
min_samples_split: 2

1\
3\
**2**\
7

Q4. Launch the tracking server locally
Now we want to manage the entire lifecycle of our ML model. In this step, you'll need to launch a tracking server. This way we will also have access to the model registry.

Your task is to:

launch the tracking server on your local machine,
select a SQLite db for the backend store and a folder called artifacts for the artifacts store.
You should keep the tracking server running to work on the next two exercises that use the server.

In addition to backend-store-uri, what else do you need to pass to properly configure the server?

**default-artifact-root**\
serve-artifacts\
artifacts-only\
artifacts-destination

In [24]:
import subprocess

In [25]:
import pandas as pd

from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error

import mlflow

In [26]:
mlflow.set_tracking_uri("file:///mlflow.db")
mlflow.set_experiment("nyc-green-taxi-experiment")


<Experiment: artifact_location='file:///mlflow.db/486535021286036824', creation_time=1748155104342, experiment_id='486535021286036824', last_update_time=1748155104342, lifecycle_stage='active', name='nyc-green-taxi-experiment', tags={}>

In [28]:
with mlflow.start_run():

    mlflow.set_tag("developer", "matt")

    mlflow.log_param("data-path", data_path_output)

    process = subprocess.run(["python","/home/mlops-zoomcamp/cohorts/2025/02-experiment-tracking/homework/train.py"\
                          , "--data_path", "/home/mlops-zoomcamp/notebooks/data/output/"], stdout = subprocess.PIPE, stderr = subprocess.PIPE)
    rmse = process.stdout.decode("utf-8").splitlines()[0].split(":")[1].strip()

    mlflow.log_metric("rmse", rmse)

    mlflow.log_artifact(local_path="/home/mlops-zoomcamp/notebooks/models/random_forest_reg.bin", artifact_path="models_pickle")

Q5. Tune model hyperparameters
Now let's try to reduce the validation error by tuning the hyperparameters of the RandomForestRegressor using hyperopt. We have prepared the script hpo.py for this exercise.

Your task is to modify the script hpo.py and make sure that the validation RMSE is logged to the tracking server for each run of the hyperparameter optimization (you will need to add a few lines of code to the objective function) and run the script without passing any parameters.

After that, open UI and explore the runs from the experiment called random-forest-hyperopt to answer the question below.

Note: Don't use autologging for this exercise.

The idea is to just log the information that you need to answer the question below, including:

the list of hyperparameters that are passed to the objective function during the optimization,
the RMSE obtained on the validation set (February 2023 data).
What's the best validation RMSE that you got?

4.817\
5.335\
5.818\
6.336